### PR TITLE
[FLINK-33823] Make PlannerQueryOperation SQL serializable

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

Make PlannerQueryOperation SQL serializable

## Verifying this change

Added tests in:
* org.apache.flink.table.api.QueryOperationSqlSerializationTest
* org.apache.flink.table.api.QueryOperationSqlExecutionTest

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/PlannerQueryOperation.java [hunk 1]
```diff
@@ -72,6 +75,13 @@ public String asSummaryString() {
                 "PlannerNode", Collections.emptyMap(), getChildren(), Operation::asSummaryString);
     }
 
+    @Override
+    public String asSerializableString() {
+        final RelToSqlConverter relToSqlConverter = new RelToSqlConverter(AnsiSqlDialect.DEFAULT);
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/PlannerQueryOperation.java:80]: In any case we need try-catch around it. Or will e.g. Table API queries that take the traditional route properly serialize?

[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/PlannerQueryOperation.java:80]: We should not expose Calcite exceptions in this case.

[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/PlannerQueryOperation.java:80]: > Since PlannerQueryOperation comes right out of the planner we can also pass some context into it. Don't we have some RelToSqlConverter or Flink dialect somewhere?

I did not look thoroughly enough. Actually we do have https://github.com/apache/flink/blob/78aa02f106cfe261200d4efce3c09d2a2882ebec/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/converters/SqlNodeConverter.java#L91

I'll use this to create the SQL string in `PlannerQueryOperation`.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/PlannerQueryOperation.java [hunk 2]
```diff
@@ -72,6 +77,15 @@ public String asSummaryString() {
                 "PlannerNode", Collections.emptyMap(), getChildren(), Operation::asSummaryString);
     }
 
+    @Override
+    public String asSerializableString() {
+        try {
+            return toSqlString.get();
+        } catch (Exception e) {
+            throw new TableException("Could not convert Calcite tree to a SQL string", e);
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/PlannerQueryOperation.java:85]: I would rephrase to `Given plan is not serializable into SQL` or so. Not mention Calicte.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/utils/TableTestBase.scala [hunk 3]
```diff
@@ -1646,6 +1650,28 @@ object TableTestUtil {
     planner.translateToRel(modifyOperation)
   }
 
+  def toQuotedSqlString(relNode: RelNode, tEnv: TableEnvironment): String = {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/utils/TableTestBase.scala:1653]: Who is calling this method?

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the update @dawidwys. The new implementation looks safer to me. I just had 2 comments left.

**Review by twalthr — APPROVED** [review]
LGTM
